client/log_client.go

```diff
@@ -215,34 +215,26 @@ func (c *LogClient) GetRoot() *types.LogRootV1 {
 	return &ret
 }
 
-// updateRootIfNewer updates the current trusted root if the new root is newer
-// than the currently trusted root and its tree size is larger.
-func (c *LogClient) updateRootIfNewer(newTrusted *types.LogRootV1) bool {
-	c.rootLock.Lock()
-	defer c.rootLock.Unlock()
-
-	currentlyTrusted := &c.root
-	if newTrusted.TimestampNanos > currentlyTrusted.TimestampNanos &&
-		newTrusted.TreeSize >= currentlyTrusted.TreeSize {
-		// Take a copy of the new trusted root in order to prevent clients from modifying it.
-		c.root = *newTrusted
-		return true
-	}
-
-	return false
-}
-
 // UpdateRoot retrieves the current SignedLogRoot, verifying it against roots this client has
 // seen in the past, and updating the currently trusted root if the new root verifies, and is
 // newer than the currently trusted root.
 func (c *LogClient) UpdateRoot(ctx context.Context) (*types.LogRootV1, error) {
 	currentlyTrusted := c.GetRoot()
-	newTrusted, err := c.getLatestRoot(ctx, currentlyTrusted)
+	newTrusted, err := c.getAndVerifyLatestRoot(ctx, currentlyTrusted)
 	if err != nil {
 		return nil, err
 	}
 
-	if c.updateRootIfNewer(newTrusted) {
+	c.rootLock.Lock()
+	defer c.rootLock.Unlock()
+
+	// Get the root again, since it might have been updated by a different thread
```

Comment:
I like your thinking in delaying the lock, however, let me elaborate on what needs to be verified. 

In addition to signature checks, `getAndVerifyLatestRoot` also verifies that there is a consistency proof between `currentlyTrusted` and `newTrusted` to form a continuous chain of checks for each state of `c.root`. 

Consider the following example of a fork where `B'` indicates a forked state:  
Both the transitions `A -> B'` and `A->C` have valid transition proofs.  `B' -> C` is invalid.
 
```
A -    - > C
  \       
    B'  
```

If a concurrent `UpdateRoot` verifies and sets `A -> B'` while we are verifying `A -> C`, `c.root` will go through an invalid state transition `A->B'->C`.  

We could either lock `UpdateRoot` so that it can't be called concurrently, or upon detecting that `currentlyTrusted` has changed, fetch a new proof for the new `currentlyTrusted`.  What do you think?

Reply:
Makes sense. Thank you for the explanation. I added an update lock to prevent updates from happening concurrently.